02-intermediate-ruby/api-post.md

```diff
@@ -0,0 +1,282 @@
+
+Next we need to see how Slack takes in messages to post into a channel.  Take a look at the [chat.postMessage](https://api.slack.com/methods/chat.postMessage) page in the Slack API documentation.  
+
+**Questions**:  
+
+- What is the URL for this Api endpoint?
+- What arguments does the `chat.postMessage` endpoint accept?  Are any mandatory?
+- What types of errors could the API return?
+
+
+### From Postman
+
+You can make a `POST` request from Postman in a similar fashion to `GET` requests.  Open [Postman](https://www.getpostman.com/downloads/).
+
+1.  First set the HTTP Method and URL in Postman, the URL should be:  `https://slack.com/api/chat.postMessage` and the method should be `POST`.
+1.  Next go to the `body` tab in postman, select `x-www-form-urlencoded` and enter the following arguments and values:
+    * token
+    * text
+    * channel
+
+Postman should look like this:
+
+![screenshot of postman making postMessage request](images/postman-slack-postMessage.png)
+
+Now hit `send`
+
+**Questions**
+
+- What JSON is returned by Slack?  What status code is returned?
+- Try submitting an invalid token, what happens?
+
+You've now successfully posted to Slack from Postman!
+
+#### Thinking About Testing
+
+**With a partner** discuss what are the things to test when sending messages to slack.  Consider the elements of the body section as arguments to a method, how can they vary.
+
+- What is the nominal case?
+- Write out all the edge-cases you can think of.  What could go wrong when sending messages to Slack?
+  - Try them out in Postman to see the results.
+
+### Using HTTParty - Setting Up
+
+In this lesson we will build a method for sending messages to Slack in a TDD fashion.  Start by forking and cloning the [Ada Gold slack-cli](https://github.com/AdaGold/slack-cli) repository.  
+
+### What Do We Have?
+
+In this repository we have a `Rakefile` to run our tests with, a `run_chatstream.rb` file to run our finished application, and a `specs` folder with the following files?
+
+- `spec_helper.rb`
+- `slack_api_wrapper_spec.rb`
+
+The `lib` directory also has a `slack_api_wrapper.rb` file with the following code inside:
+
+```ruby
+# slack_api_wrapper.rb
+require 'httparty'
+
+class SlackApiWrapper
+  BASE_URL = 'https://slack.com/api/'
+
+  # Code goes here
+
+end
+```
+
+In this exercise we will build some tests for a `send_msg` method in the `SlackApiWrapper` class and then build out the functionality to satisfy the tests.  
+
+### Our First Test - Nominal Case
+
+For our first test we will build a test case for sending a message to Slack successfully.  We will need to set up the testing first.  Add the following to `specs/spec_helper.rb`
+
+```ruby
+# spec_helper.rb
+require 'dotenv'
+Dotenv.load
```

Comment:
Do we need this for a non-rails project?

Reply:
It might be worth calling out that if we ever do this in a Rails project, we won't need to include this